apm-agent-plugins/apm-spring-webflux/apm-spring-webflux-plugin/src/main/java/co/elastic/apm/agent/springwebflux/TransactionAwareSubscriber.java

```diff
@@ -218,7 +181,26 @@ private void doExit(boolean discard, String method, @Nullable Transaction transa
             }
             transactionMap.remove(this);
         }
+    }
+
+    public void cancelTransaction() {
+        Transaction transaction = getTransaction();
+        debugTrace(true, "cancelTransaction", transaction);
+        try {
+            if (transaction == null) {
+                return;
+            }
+
+            WebfluxHelper.endTransaction(null, transaction, exchange);
+
+            // not 100% sure to understand why it's required, but without it recycling reference count
+            // is always off by one when transaction get cancelled
```

Comment:
This is something we must figure out. Could it be that the transaction was activated but not deactivated in this scenario?

Regardless to this specific (I think), this class stores transaction references in the `transactionMap`, but we do not reflect that in the reference counting. The ref count needs to be incremented when the transaction is added to the map and decremented when removed.

Reply:
After a bit much more investigation following our discussion today, this extra decrement is required because the transaction is not being properly de-activated (which is expected).

- We can't simply limit activation to the wrapped method bodies (`onSubscribe`, `onNext`, `onComplete`, `onError`) like with the reactor plugin because otherwise we have failing tests, for example trying to set a custom transaction name with annotated controllers.
- as it seems to already work better like this, I would suggest to keep it as-is and revisit this later if that becomes an issue.
- there might be other ways to cover the "set custom transaction name" test, for example by wrapping/instrumenting, but that would require some further investigation and investment.
- I think it could have been a bit easier if the "current transaction" was being accessible through Webflux context storage instead of relying on current thread state and activation/deactivation events (but that would be a major change).

As `transactionMap` is an instance of `SpanConcurrentHashMap`, the reference count is already managed by the map implementation.

Reply:
> there might be other ways to cover the "set custom transaction name" test, for example by wrapping/instrumenting, but that would require some further investigation and investment.

If that allows us to get rid of the unsafe pattern of activating a span in one place and deactivating it somewhere else in favor of activating/deactivating in something resembling a try/finally block, I think that'd be well worth it.

> I think it could have been a bit easier if the "current transaction" was being accessible through Webflux context storage instead of relying on current thread state and activation/deactivation events

We have to explicitly activate so that automatically created spans for HTTP clients connect to the WebFlux transaction, for example. But maybe we can internally rely on the context: For example, before the controllers are called, we can instrument that part of the dispatching, get the transaction from the WebFlux context and activate it.